modules/metrics-probe/filterx/func-update-metric.c

```diff
@@ -22,3 +22,196 @@
  */
 
 #include "func-update-metric.h"
+#include "filterx/filterx-metrics.h"
+#include "filterx/expr-literal.h"
+#include "filterx/filterx-eval.h"
+#include "filterx/object-primitive.h"
+#include "stats/stats.h"
+
+#define FILTERX_FUNC_UPDATE_METRIC_USAGE "update_metric(\"key\", labels={\"key\": \"value\"}, increment=1, level=0)"
+
+typedef struct FilterXFunctionUpdateMetric_
+{
+  FilterXFunction super;
+  FilterXMetrics *metrics;
+  gint level;
+
+  struct
+  {
+    FilterXExpr *expr;
+    gint64 value;
+  } increment;
+} FilterXFunctionUpdateMetric;
+
+static gboolean
+_get_increment(FilterXFunctionUpdateMetric *self, gint64 *increment)
+{
+  if (!self->increment.expr)
+    {
+      *increment = self->increment.value;
+      return TRUE;
+    }
+
+  FilterXObject *increment_obj = filterx_expr_eval_typed(self->increment.expr);
+  if (!increment_obj)
+    return FALSE;
+
+  gboolean success = filterx_integer_unwrap(increment_obj, increment);
+  if (!success)
+    filterx_eval_push_error("metric increment must be an integer", self->increment.expr, increment_obj);
+
+  filterx_object_unref(increment_obj);
+  return success;
+}
+
+static FilterXObject *
+_eval(FilterXExpr *s)
+{
+  FilterXFunctionUpdateMetric *self = (FilterXFunctionUpdateMetric *) s;
+
+  gboolean success = FALSE;
+
+  gint64 increment;
+  if (!_get_increment(self, &increment))
+    goto exit;
+
+  StatsCounterItem *counter;
+  if (!filterx_metrics_get_stats_counter(self->metrics, &counter))
+    goto exit;
+
+  stats_counter_add(counter, increment);
+  success = TRUE;
+
+exit:
+  if (!success)
+    {
+      msg_error("FilterX: Failed to process update_metric()", filterx_format_last_error());
```

Comment:
I think it is a bad idea to start generating a message in these cases. I think it'd be much better to add an "error" counter along with all update_metric() counters and increment that every time something fails.

With that we could even add limits on how many counters we would be adding in case there's a cardinality explosion.

it does not need to be done now, just an idea.

Reply:
I agree, changed the log level to debug in a new commit. We can introduce a counter later.